src/python_rwpm_parser/parser/syntax.py

```diff
@@ -0,0 +1,232 @@
+        :param json_content: Dictionary containing collection's JSON
+        :type json_content: Dict
+
+        :return: List of Link objects
+        :rtype: List[Link]
+        """
+        self._logger.debug('Started parsing links')
+
+        if isinstance(json_content, dict):
+            links = json_content[self.LINKS]
+        elif isinstance(json_content, list):
+            links = json_content
+        else:
+            raise ParsingError('Links list has an incorrect format')
+
+        link_objects = []
+
+        for link in links:
+            self._logger.debug('Started parsing link {0}'.format(link))
+
+            link_object = self._parse_object(link, Link)
+            link_objects.append(link_object)
+
+            self._logger.debug('Finished parsing link {0}: {1}'.format(link, link_object))
+
+        self._logger.debug('Finished parsing links: {0}'.format(link_objects))
+
+        return link_objects
+
+    def _parse_collection(self, json_content, collection=None, role=None):
+        """Parses collection's JSON into a Collection object
+
+        :param json_content: Dictionary containing collection's JSON
+        :type json_content: Dict
+
+        :param collection: Optional object containing collection object. If missing, a new Collection will be created
+        :type collection: Optional[Collection]
+
+        :param role: Optional CollectionRole object (empty when the method is called for a manifest)
+        :type role: CollectionRole
+
+        :return: Collection object
+        :rtype: CompactCollection
+        """
+        self._logger.debug('Started parsing collection {0} with role {1}'.format(collection, role))
+
+        if collection is None:
+            collection = Collection(role)
+
+        if not role or not role.compact:
+            links = self._parse_links(json_content)
+            collection.links.extend(links)
+
+            metadata = self._parse_metadata(json_content)
+            collection.metadata = metadata
+
+            sub_collection_keys = set(json_content.keys()) - {'@context', 'metadata', 'links'}
+
+            self._logger.debug('There are following potential sub-collections: {0}'.format(sub_collection_keys))
+
+            for collection_role in self._collection_roles_registry:
+                if collection_role.key in sub_collection_keys:
+                    sub_collection = self._parse_collection(
+                        json_content[collection_role.key], collection=None, role=collection_role)
+
+                    collection.sub_collections.append(sub_collection)
+
+                    sub_collection_keys.remove(collection_role.key)
+                elif collection_role.required:
+                    raise ParsingError('Missing required sub-collection {0}'.format(collection_role.key))
+
+            if sub_collection_keys:
+                raise ParsingError(
+                    'The following sub-collection keys are not registered: {0}'.format(sub_collection_keys))
```

Comment:
I don't think this should be a parsing error, since RWPM allows extension roles (though it says "Extensions that are not registered in the registry must use a URI for their role.") I suggest treating these leftover keys as CollectionRoles with `compact=False` and `require=False`.

Then, when validating the collection, issue a warning if a collection has a role that is neither in (our model of) the registry nor a URI.

Reply:
Currently the parser raises an exception and halts on the first failure. Would you like to have it to return a list of errors, warnings instead?

Reply:
I was saying that this shouldn't even be considered an error. I think you're saying that it should be considered an error, but that an error doesn't necessarily need to stop the parsing process.

So, two questions to resolve: is this actually an error? And should a problem like this necessarily result in an exception?

First, I don't think we can say for sure that this particular thing is an error, because we only know the contents of the RWPM role registry at the time of the release of the software. I don't want to have to put out an immediate upgrade if a role is added to the registry so we can parse the new role without an error.

That said, if an "error" in this case just means something is added to a list, then it's liveable, and I'd be okay with returning a 2-tuple of (errors, warnings).

Reply:
I removed this code and the parser doesn't raise an exception in the case of unknown collections